File: Cuda_kernels/heat_equation_solver/src/HeatEquationSolver.cpp
#include "HeatEquationSolver.h"
#include <iostream>
#include <omp.h>
#include <nvtx3/nvToolsExt.h>  // NVTX header for profiling

void cpu_heat_equation_solver(int N, int num_steps, float *grid, float alpha, float dt, float dx2, float dy2, int num_threads) {
    float *new_grid = new float[N * N];
    omp_set_num_threads(num_threads);

    // Add NVTX marker for the main loop
    nvtxRangeId_t cpu_range = nvtxRangeStartA("CPU Solver");

    for (int step = 0; step < num_steps; ++step) {
        #pragma omp parallel for collapse(2)
        for (int i = 1; i < N - 1; ++i) {
            for (int j = 1; j < N - 1; ++j) {
                int index = i * N + j;
                float temp = grid[index];
                float temp_left = grid[i * N + (j - 1)];
                float temp_right = grid[i * N + (j + 1)];
                float temp_top = grid[(i - 1) * N + j];
                float temp_bottom = grid[(i + 1) * N + j];
                new_grid[index] = temp + alpha * dt * ((temp_left - 2.0f * temp + temp_right) / dx2 +
                                                       (temp_top - 2.0f * temp + temp_bottom) / dy2);
            }
        }
        std::swap(grid, new_grid);
    }

    nvtxRangeEnd(cpu_range);  // End NVTX marker for the CPU solver
    delete[] new_grid;
}

File: Cuda_kernels/heat_equation_solver/src/HeatEquationSolver.py
import numpy as np
from heat_solver import cpu_heat_solver, gpu_heat_solver

class HeatEquationSolver:
    def __init__(self, N, block_thread_x=16, block_thread_y=16):
        self.N = N
        self.block_thread_x = block_thread_x
        self.block_thread_y = block_thread_y
        self.alpha = 0.5
        self.dx = self.dy = 0.01
        self.dt = self.dx * self.dy / (2.0 * self.alpha * (self.dx * self.dx + self.dy * self.dy))
        self.dx2 = self.dx * self.dx
        self.dy2 = self.dy * self.dy

    def initialize_grid(self):
        grid = np.full((self.N, self.N), 25.0, dtype=np.float32)
        grid[self.N // 2, self.N // 2] = 100.0
        return grid

    def solve_cpu(self, num_steps, num_threads=None):
        grid = self.initialize_grid()
        cpu_heat_solver(self.N, num_steps, self.alpha, self.dt, self.dx2, self.dy2, num_threads or 0, grid)
        return grid

    def solve_gpu(self, num_steps):
        grid = self.initialize_grid()
        gpu_heat_solver(self.N, num_steps, self.alpha, self.dt, self.dx2, self.dy2, 
                        self.block_thread_x, self.block_thread_y, grid)
        return grid